src/test/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManagerTest.java

```diff
@@ -0,0 +1,89 @@
+
+public class UserAccountManagerTest {
+
+    @ParameterizedTest
+    @MethodSource("namesToTest")
+    public void nameValidation(boolean expectedIsValid, String name) {
+        boolean actualIsValid = isUserNameValid(name);
+        assertEquals(expectedIsValid, actualIsValid);
+    }
+
+    private static Stream<Arguments> namesToTest() {
+        return Stream.of(
+                Arguments.of(true, "testname"), // Lowercase is valid
+                Arguments.of(true, "TESTNAME"), // Uppercase is valid
+                Arguments.of(true, "TestName"), // Mixture of cases is valid
+                Arguments.of(true, "TestName123"), // Numbers are valid
+                Arguments.of(true, "Test_Name"), // Underscores are valid
+                Arguments.of(true, "Test Name"), // Double-barrelled names with spaces are valid
+                Arguments.of(true, "Test-Name"), // Double-barrelled names with hyphens are valid
+                Arguments.of(true, "O'Name"), // Apostraphes are valid
+                Arguments.of(false, null), // Null is not valid
+                Arguments.of(false, ""), // Empty string is not valid
+                // Names exceeding the maximum length of 255 characters are not valid
+                Arguments.of(false, "TestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestNameTestName"),
+                // Other special characters are not permitted
+                Arguments.of(false, "Test!Name"),
+                Arguments.of(false, "Test?Name"),
+                Arguments.of(false, "Test#Name"),
+                Arguments.of(false, "Test.Name"),
+                Arguments.of(false, "Test,Name"),
+                Arguments.of(false, "Test*Name"),
+                Arguments.of(false, "Test<Name"),
+                Arguments.of(false, "Test>Name"),
+                Arguments.of(false, "Test:Name"),
+                Arguments.of(false, "Test;Name"),
+                Arguments.of(false, "Test/Name")
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("emailsToTest")
+    public void emailValidation(boolean expectedIsValid, String email) {
+        boolean actualIsValid = isEmailValid(email);
+        assertEquals(expectedIsValid, actualIsValid);
+    }
+
+    private static Stream<Arguments> emailsToTest() {
+        return Stream.of(
+                Arguments.of(true, "testemail@test.com"), // Standard email format is valid
+                // Alternate pattern for twitter|google|facebook is valid
```

Comment:
Just out of curiosity, how come this is allowed? I understand that this predates your changes, but I've never seen this before. Does this have anything to do with OAuth/OpenID Connect?

Reply:
I'm _assuming_ this is something to do with third party authentication, yes. The authentication process supports a number of alternatives in additional to the local version and all three of those domains are listed. Haven't found why they have an address format like that yet though...
```
public enum AuthenticationProvider {
	GOOGLE, FACEBOOK, TWITTER, RAVEN, TEST, SEGUE, RASPBERRYPI
}
```

Reply:
It appends '-twitter' in TwitterAuthenticator.getUserInfo(). Same for the others. There's some rather hacky stuff going on in there.

Reply:
Going off the discussion in standup this morning, it sounds like we're not taking any further action with regards to the third-party authentication in the ticket so this should be ready for review again?

Reply:
Could you humour me and check whether there is a flag anywhere in the db schema that seems to tell you what 3rd party IdP is being used. I'm going to guess 'no'. It doesn't make much odds to this ticket really, but it would be useful information.

Reply:
OK, I've dug a bit further. The assumption that email is not null is baked into various parts of the codebase. At the same time, 3rd party IdPs may not return an email (the user may choose to withhold it for example). The authenticator classes replace email in the user object with provider userid appended with the provider string where this is the case. It's not perfect, but it's outside the scope of this ticket to change it.